tensorflow/src/TensorFlow/Internal/FFI.hs

```diff
@@ -75,25 +76,26 @@ data TensorData = TensorData
 
 -- | Runs the given action after creating a session with options
 -- populated by the given optionSetter.
-withSession :: (Raw.SessionOptions -> IO ())
-            -> ((IO () -> IO ()) -> Raw.Session -> IO a)
+withSession :: (MonadIO m, MonadMask m)
+            => (Raw.SessionOptions -> IO ())
+            -> ((IO () -> IO ()) -> Raw.Session -> m a)
             -- ^ The action can spawn concurrent tasks which will
             -- be canceled before withSession returns.
-            -> IO a
+            -> m a
 withSession optionSetter action = do
-    drain <- newMVar []
+    drain <- liftIO $ newMVar []
     let cleanup s =
         -- Closes the session to nudge the pending run calls to fail and exit.
             finally (checkStatus (Raw.closeSession s)) $ do
                 runners <- takeMVar drain
                 -- Collects all runners before deleting the session.
                 mapM_ shutDownRunner runners
                 checkStatus (Raw.deleteSession s)
-    bracket Raw.newSessionOptions Raw.deleteSessionOptions $ \options -> do
-        optionSetter options
+    bracket (liftIO $ Raw.newSessionOptions) (liftIO . Raw.deleteSessionOptions) $ \options -> do
+        liftIO $ optionSetter options
         bracket
-            (checkStatus (Raw.newSession options))
-            cleanup
+            (liftIO $ checkStatus (Raw.newSession options))
+            (liftIO . cleanup)
```

Comment:
5 liftIO in 4 lines feels like an overkill. Can this code be refactored a bit to reduce verbiage? It feels almost like you can pull this outside if it were possible to "unliftIO" the line below. Not that I'm advocating adding a dependency on 
https://hackage.haskell.org/package/unliftio ...

Reply:
Here is one idea:
```haskell
    let bracketIO x y = bracket (liftIO x) (liftIO . y)
    bracketIO Raw.newSessionOptions Raw.deleteSessionOptions $ \options -> do
        bracketIO
            (optionSetter options >> checkStatus (Raw.newSession options))
            cleanup
            (action (asyncCollector drain))
```

Reply:
Yeah, this is much nicer. Let's make it so :)